### ISSUE TITLE
Support for RSocket

### ISSUE THREAD

**Issue by Scisci75** [quarkusio/quarkus#1997]
Hi, 
have you thought about adding support for the RSocket protocol?
I think it's a good replacement/alternative for Rest communications.
Thanks
Mirko

**Comment by kamilgregorczyk** [comment]
I agree, especially with rsocket-rpc-java which is amazing and fully reactive, I think it would fit quarkus perfectly

**Comment by hantsy** [comment]
Like RSocket support in Spring, I think there are a few things need to be done if considering RSocket as an alternative of REST.
1. Standalone lightweight Netty based embedded RSocket server via TCP/Websocket, or add Rsocket Support to an existing Netty server(serves Rest over HTTP, and TCP/Websocket for Rsocket at the same time).
2. RSocket client and server-side programming abstraction and communication, and provides `RSocketRequestor` some like helpers in client side , and reuse the Microprofile Reactvie Messages APIs to simplify the programming in server side, like what is done in Spring, which use Spring Messaging to hide the complexity of RSocket. 
3. Extract Quarkus security core outside of certain protocols, make it not depends on HTTP, Websocket, Rosocket,  gRpc, etc, add Quarkus security support for RSocket(connection setup and request guard).

**Comment by tlvenn** [comment]
Shouldn't the issue relabeled as 'kind/extension-proposal' / 'kind/new-feature' @gsmet ? It seems weird to tag it as a question. Thanks

**Comment by oleksiylukin** [comment]
RSocket extension would be good solution for fast and reliable bi-directional communication between microservices. RSocket protocol can run over simple TCP socket or even over UDP with minimal overhead and it is much faster then any HTTP/1.1 or HTTP/2 based websocket channels

**Comment by nimo23** [comment]
Has RSocket (unlike WebSocket) built-in support for re-connections?

**Comment by tlvenn** [comment]
@nimo23 yes -> https://rsocket.io/docs/Protocol#resuming-operation

**Comment by cyberquarks** [comment]
Hi, is there a plan to support rsocket with Quarkus?

**Comment by Andro999b** [comment]
Any plans for this?

**Comment by alukin** [comment]
It would be great to have quick reactive communications channels between microservices. REST is good but it is not duplex; Websockets are not independent and bound to servelts. So if you need to pass messages between microservices you have to use some "monster" extension with a central messaging service or invent the weel on top of netty for the simple P2P communication channel.
So here comes rshocket to solve the problem.

**Comment by dufoli** [comment]
I have start to implement an extension here : https://github.com/dufoli/quarkus-rsocket

**Comment by dufoli** [comment]
@cescoffier I have started to work on an extension.
1) do you plan to integrate it on quarkus on quarkiverse ?
2) Which microprofile annotation I can used Path for routing seems good but to split mode between fire and forget, channel, request and response, ...
Anyone has ever think on design to go in right direction ?

**Comment by hantsy** [comment]
Spring RSocket reused the annotations in spring-messaging,`@Controller`, `@MessagingMapping` , etc.

Quarkus create a new series of annotations `@Route` like annotations for vertex, but no united handlers for messaging cases.

If possible I think provide a simple fluent  pragmatic APIs to assemble routes firstly. @dufoli

**Comment by hantsy** [comment]
For content type, you should consider metadata and data payload respectively, they are separated in RSocket.

**Comment by hantsy** [comment]
And for the message encoders or decoders(Spring has similar abstraction), not sure Quarkus has an united config can be applied to different case, such as HTTP, MP messaging, GRPC(probuf), RSocket. 

Quarkus itself provides a simple  Jaskson ObjectMapperCustomizer and jsonb JsonBConfigCustomizer to custom ObjectMapper and Jsonb.

**Comment by dufoli** [comment]
io.quarkus.vertx.web.Route is inside vertx extension, I do not want to depend on it whereas I use netty and route work with method.
In the other hand, Micro profile is more generic.
so 
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#pathItemObject
Path annotation is better for routing

step 1

routing will be done to RSocket implementation with Path.
If no path annotation it will be just the directly used without routing for all request.
 
```
@Path("/foo")
public class MysSocket implements RSocket {
@override
public Mono<Void> fireAndForget(Payload payload) {
    return null;
  }

 @override
  public Mono<Payload> requestResponse(Payload payload) {
    return null;
  }

@override
  public Flux<Payload> requestStream(Payload payload) {
    return null;
  }

@override
  public Flux<Payload> requestChannel(Publisher<Payload> payloads) {
    return null;
  }
}
```

Step 2:
switch from payload to object for that I will used the consumes and Produces with MediaType annotation
```
  @Consumes({MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML})
  @Produces({MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML})
```

Step 3: 
to be determined but handle mode and transport with annotation can be a good point too.

any feedback welcome before going further ;-)

**Comment by cescoffier** [comment]
+1 for quarkiverse

About the annotation, `@Route` is quite HTTP specific. What about the JAX-RS `@Path`? 

For the RSocket class, we would need the Mutiny variant.

**Comment by hantsy** [comment]
I do not think **Jaxrs** `@Path` is a good option. Personally, I hope Quakus will provide some common abstract APIs for messaging handling(for MP messaging, AMQP, JMS, WebSocket/STOMP, RSocket, etc). 

I would use an MP `ReactiveStreams` util like class to build the RSocket facilities now.

eg.  `RSocketBuilders` to create client requestor and start the server, `RSocketRequestor` to send request,  `RSocketRouteHandlers` to receive the data.  RSocket is bidirectional, so there maybe exists an `RSocketRequestor` for server-client request sending.

**Comment by yschimke** [comment]
@dufoli This looks great.  Just wanted to point out we have a public demo server at wss://rsocket-demo-1.appspot.com/rsocket with support for a few spring routes if you want to test against rsocket-java.  Likewise there is http://github.com/making/rsc (java) and http://github.com/rsocket/rsocket-cli (kotlin) clients if you want to test a known working client against your server.

**Comment by dufoli** [comment]
@gastaldi can you create a quarkiverse repo please ?

**Comment by gastaldi** [comment]
@dufoli Repository created in https://github.com/quarkiverse/quarkus-rsocket. I made you a maintainer of that repository too 😉 

Thank you!

**Comment by dufoli** [comment]
I have move my code to quarkiverse repo and add few issue to have a global roadmap.
If people want to contribute or request features please used the quarkiverse repo ;-)
